File: java/pe/src/ktn/pe/Pe8.java
package ktn.pe;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.logging.Logger;

public class Pe8 implements Pe {
    private static final Logger logger = Logger.getLogger(Pe8.class.getName());
    private static final String defaultFilename = "../res/pe8.txt";
    
    private String fname = defaultFilename;
    private int digits = 5;

    private String loadFile(File file) {
        if (file == null) {
            return null;
        }

        StringBuffer buf = new StringBuffer();
        try (FileInputStream fis = new FileInputStream(file);
                InputStreamReader isr = new InputStreamReader(fis, "UTF-8");
                BufferedReader br = new BufferedReader(isr)) {
            String line;
            while ((line = br.readLine()) != null) {
                buf.append(line);
            }
        } catch (IOException e) {
            logger.severe(e.toString());
        }

        return buf.toString();
    }

    public int product(String numbers, int n) {
        int m = 1;
        if (numbers == null) {
            return m;
        }

        int t;
        int length = numbers.length();
        for (int i = 0; i < length - n + 1; ++i) {
            t = 1;
            for (int j = 0; j < n; ++j) {
                t *= numbers.charAt(i + j) - '0';
            }
            if (m < t) {
                m = t;
            }
        }

        return m;
    }

    public int pe8() {
        return pe8(fname, digits);
    }
    
    public int pe8(int n) {
        return pe8(fname, n);
    }

    public int pe8(String fname, int n) {
        File file = new File(fname);
        String numbers = loadFile(file);
        logger.fine(numbers);

        return product(numbers, n);
    }

    @Override
    public void setArgs(String[] args) {
        if (args == null) {
            return;
        }

        if (args.length > 0) {
            try {
                digits = Integer.parseInt(args[0]);
            } catch (NumberFormatException e) {
                //
            }
        }
        if (args.length > 1) {
            fname = args[1];
        }
    }

    @Override
    public void solve() {
        System.out.println(PeUtils.format(this, pe8()));
    }

    @Override
    public void run() {
        solve();
    }

    @Override
    public int getProblemNumber() {
        return 8;
    }

    public String getFname() {
        return fname;
    }

    public void setFname(String fname) {
        this.fname = fname;
    }

    public int getDigits() {
        return digits;
    }

    public void setDigits(int digits) {
        this.digits = digits;
    }

}

File: kt/pe/src/main/kotlin/ktn/pe/Pe8.kt
package ktn.pe

import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.util.logging.Logger

class Pe8 : Pe {

    var fname = DEFAULT_FILENAME
    private var digits = 5

    override val problemNumber: Int
        get() = 8

    private fun loadFile(file: File): String {
        val buf = StringBuffer()
        try {
            FileInputStream(file).use({ fis ->
                InputStreamReader(fis, "UTF-8").use({ isr ->
                    BufferedReader(isr).use({ br ->
                        var line: String?
                        while (true) {
                            line = br.readLine()
                            if (line == null) { break; }
                            buf.append(line)
                        }
                    })
                })
            })
        } catch (e: IOException) {
            logger.severe(e.toString())
        }

        return buf.toString()
    }

    fun product(numbers: String, n: Int): Int {
        var m = 1
        var t: Int
        val length = numbers.length
        for (i in 0 until length - n + 1) {
            t = 1
            for (j in 0 until n) {
                t *= numbers[i + j] - '0'
            }
            if (m < t) {
                m = t
            }
        }

        return m
    }

    fun pe8(n: Int): Int {
        return pe8(fname, n)
    }

    @JvmOverloads
    fun pe8(fname: String = DEFAULT_FILENAME, n: Int = digits): Int {
        val file = File(fname)
        val numbers = loadFile(file)
        logger.fine(numbers)

        return product(numbers, n)
    }

    override fun setArgs(args: Array<String>) {
        if (args.isNotEmpty()) {
            digits = args[0].toInt()
        }
        if (args.size > 1) {
            fname = args[1]
        }
    }

    override fun solve() {
        System.out.println(PeUtils.format(this, pe8()))
    }

    override fun run() {
        solve()
    }

    companion object {
        private val logger = Logger.getLogger(Pe8::class.java.name)
        private const val DEFAULT_FILENAME = "../../res/pe8.txt"
    }

}
